cDac work necessary to get the MethodTableName

davidwrighton opened the pull request:
Add new details to the RuntimeTypeSystem and Loader contracts as needed to load metadata and examine and identify all of the kinds of type that the CoreCLR type system can represent

Add a type name generator based on a line by line copy from the CoreCLR codebase

Add an Ecma335 metadata parser which is pure safe managed code, and is capable of loading metadata that is not structured as a single array. This implementation is heavily based on the dnmd implementation.

Provide impelementations and documentation for all of the new contracts except for the RW metadata one. (Its rarely needed, so we can add it later)

Enhance the target infrastructure to better handle various forms of arrays, and contracts which do math based on target pointer sizes.

Contributes to https://github.com/dotnet/runtime/issues/99302

davidwrighton committed:
Progress towards GetMethodTableName

davidwrighton committed:
Move metadata reader into helpers
Add implementations for changes to RuntimeTypeSystem contract
Add SOSDacInterface call into cDac for GetMethodTableName

davidwrighton committed:
Simplify storage of saved metadata copy for reflection emit scenarios
- Move pointer to Module class
- Replace use of SBuffer abstraction with a simple counted byte memory block
Add metadata details to Loader contract
Add Metadata helper api for use by contracts within the cDAC (and possibly clients of cDAC too)

davidwrighton committed:
Fix bugs found when trying to actually use this logic
1. It works now!

davidwrighton committed:
Update contract documentation

davidwrighton committed:
Merge branch 'main' into cdac-methodtable-name

davidwrighton commented:
There are a few remaining TODOs ... notably, I haven't implemented sorted lookups, and the ecma api is currently only low level token and constant access, but these could be fixed in a follow up PR, or if this sits for long enough, I'll fix it here.

jkoritzinsky commented on src/native/managed/cdacreader/src/Helpers/Metadata.cs:
```diff
@@ -0,0 +1,340 @@
+    public EcmaMetadataSchema Schema { get; init; }
+
+    private ReadOnlyMemory<byte>[] _tables;
+    public ReadOnlySpan<ReadOnlyMemory<byte>> Tables => _tables;
```
Could we use `ReadOnlySequence<byte>` here? Basically, instead of having each element of the span represent the memory of a table, we could represent the underlying memory of the image as a `ReadOnlySequence<byte>` that stitches together the various components of the image.

You could use DNMD's implementation of the "save to memory" function to provide the remaining byte arrays to stitch together the separate tables and heaps into a valid ECMA-335 image represented as a `ReadOnlySequence<byte>`.

If we were to do this, we could (performance-permitting), refactor System.Reflection.Metadata to support this use case and remove the actual metadata reader implementation from here (and only have the "stitch into a sequence that represents metadata" code).

AaronRobinsonMSFT replied:
> Could we use ReadOnlySequence<byte> here? Basically, instead of having each element of the span represent the memory of a table, we could represent the underlying memory of the image as a ReadOnlySequence<byte> that stitches together the various components of the image.

My gut reaction here would be introducing this abstraction is a bit of magic I'm not sure is a win. Do we have an example I can see that shows a benefit?

davidwrighton replied:
I think that's not a bad idea, or at least, it doesn't really look like one. I don't really want to pursue it yet though.

1. This implementation should allow us to fill out the remaining metadata needs of the cDAC without all that much difficulty (other than the IMetadataImport export requirement, which System.Reflection.Metadata doesn't help with at all, in fact, I'm not sure the IMetadataImport api surface can be implemented on top of System.Reflection.Metadata).
2. I want to explore the possibility in .NET 10 of changing how we produce triage dump metadata when it becomes CDAC driven. Currently we produce an parallel data stream that has precomputed strings, but an alternative approach that might be more scalable to our real needs could be to have triage dump metadata be fully real metadata but only the exact bytes that the reader needs (not entire tables). Having a custom metadata reader makes adding that sort of hook a fair bit easier.

jkotas replied:
System.Reflection.Metadata is the managed low-level metadata reader library. I do not think we want to be growing a different one. We should refactor System.Reflection.Metadata as needed.

davidwrighton replied:
@jkotas, I don't see that as a practical near term effort. To make it able to handle the cases that come up here, would require replacing all of the memory handling of that library with something different, which I can't see happening as part of this effort until we've stabilized the capabilities of what this logic will require the metadata system to be able to process. In general, I view what I've built as a short term solution, that will either last until we replace it with dnmd, or we do a major refactor of S.R.M to make it able to operate on things that aren't a pointer to a contiguous memory blob.

AaronRobinsonMSFT commented on docs/design/datacontracts/RuntimeTypeSystem.md:
```diff
@@ -4,17 +4,64 @@ This contract is for exploring the properties of the runtime types of values on
+}
+
+
+internal enum CorElementType
```
Instead of spelling this out here, please define the the type and then comment where the values come from.

```csharp
internal enum CorElementType
{
   // Values defined in ECMA-335 - II.23.1.16 Element types used in signatures
}
```

AaronRobinsonMSFT commented on src/native/managed/cdacreader/src/Contracts/RuntimeTypeSystem_1.cs:
```diff
@@ -154,4 +157,184 @@ public uint GetTypeDefToken(MethodTableHandle methodTableHandle)
+        if (address == 0)
+            return default;
+
+        if (((ulong)address & 2) == (ulong)2)
```
```suggestion
        if (((ulong)address & 2) != 0)
```

AaronRobinsonMSFT commented on src/native/managed/cdacreader/src/Data/EEClass.cs:
```diff
@@ -15,9 +15,24 @@ public EEClass(Target target, TargetPointer address)
     public TargetPointer MethodTable { get; init; }
     public ushort NumMethods { get; init; }
     public uint CorTypeAttr { get; init; }
+    public byte InternalCorElementType { get; init; }
```
Please doc what makes this "Internal", that is generally an important distinction.

AaronRobinsonMSFT commented on src/native/managed/cdacreader/src/Helpers/Metadata.cs:
```diff
@@ -0,0 +1,340 @@
+    GenericParam = 0x2a,
+    MethodSpec = 0x2b,
+    GenericParamConstraint = 0x2c,
+    Count = 0x2d
```
```suggestion
    Count
```

AaronRobinsonMSFT reviewed (commented):
Are we adding tests for this?

davidwrighton commented on docs/design/datacontracts/RuntimeTypeSystem.md:
```diff
@@ -4,17 +4,64 @@ This contract is for exploring the properties of the runtime types of values on
+    public TargetPointer Address { get; }
+}
+
+struct TypeHandle
```
It's an opaque structure that can be used with some of the new apis. @lambdageek and I have been talking about simply using TypeHandle for all types instead of having a pair of opaque structures, and after finishing this up, I think that's the right approach. I was waiting for him to drop in on this before I made such a change.

davidwrighton commented on docs/design/datacontracts/RuntimeTypeSystem.md:
```diff
@@ -154,12 +227,39 @@ internal struct MethodTable_1
 
+Internally the contract depends on a `TypeDescHandle` structure that represents the address of something that implements the TypeDesc data descriptor.
+
+Internally the contract depends on the `TypeHandle` structure being capable of holding a TypeDescHandle or a MethodTableHandle.
```
That's why this is documented as part of version 1 of the contract. By definition the specific versions are about the particular implementation details. You'll note that the version free portion of the contract documentation does not say this at all.

Also, notably, I think we should think about changing our cdacreader dll into being constructed from 2 different dlls. 1 which is the contract and targets and such., and the second which is the legacy api surface which is what we're actually exposing from cdacreader.dll.

davidwrighton commented on src/coreclr/debug/daccess/request.cpp:
```diff
@@ -1916,7 +1916,47 @@ ClrDataAccess::GetMethodTableName(CLRDATA_ADDRESS mt, unsigned int count, _Inout
+HRESULT
+ClrDataAccess::GetMethodTableNameImpl(CLRDATA_ADDRESS mt, unsigned int count, _Inout_updates_z_(count) WCHAR *mtName, unsigned int *pNeeded)
+{
+    HRESULT hr = S_OK;
```
The impl methods are not supposed to have SOSDacEnter. that are in the wrapper method.

AaronRobinsonMSFT replied:
Oh I see now.

davidwrighton commented on src/coreclr/debug/daccess/request.cpp:
```diff
@@ -1987,7 +2027,6 @@ ClrDataAccess::GetMethodTableName(CLRDATA_ADDRESS mt, unsigned int count, _Inout
         }
     }
 
-    SOSDacLeave();
```
The impl methods are not supposed to have SOSDacEnter. that are in the wrapper method.

davidwrighton committed:
Feedback

lambdageek commented on src/native/managed/cdacreader/src/Contracts/RuntimeTypeSystem_1.cs:
```diff
@@ -154,4 +157,184 @@ public uint GetTypeDefToken(MethodTableHandle methodTableHandle)
+
+        TargetPointer dictionaryPointer = _target.ReadPointer(perInstInfo);
+
+        int numberOfGenericArgs = _target.ProcessedData.GetOrAdd<GenericsDictInfo>(genericsDictInfo).NumTypeArgs;
+        MethodTableArray instantiation = _target.ProcessedData.GetOrAdd<(TargetPointer, int), MethodTableArray>
+            ((dictionaryPointer, numberOfGenericArgs));
+
+        return instantiation.Types.AsSpan();
```
why does the key for the data include the number of generic args?  can the same `dictionaryPointer` be viewed with a different number of generic args?

davidwrighton replied:
Because just the pointer to the start of the data doesn't quite encode enough to identify the number of generic args. I'm going to restructure this so that instead of this being a general purpose `TypeHandleArray`... that it becomes a `TypeInstantiation` type, that type is defined in the contract implementation, and the pointer that makes identifies it is the `MethodTable` pointer. That should make this simpler, remove the extra key concept, remove the concern about Data contracts knowing about contracts, and make it so that repeated calls to `GetInstantiation` api are just the cost of doing data lookup, which should help us put our new design on a path to actually be *faster* than the old model.

lambdageek commented on src/native/managed/cdacreader/src/Data/MethodTableArray.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace Microsoft.Diagnostics.DataContractReader.Data;
+
+internal sealed class MethodTableArray : IData<MethodTableArray, (TargetPointer ptr, int size)>
```
this doesn't feel like it belongs in `Data`.  I've typically viewed `Data` as straightforward data-descriptor extractor.  Something we could potentially source-generate in the future.  This `MehtodTableArray` feels more like it belongs to the RuntimeTypeSystem contract proper

davidwrighton replied:
I think we need a general structure for post-processed data produced by the contracts. @lambdageek you have already done this with the _methodTables array in the `RuntimeTypeSystem` contract, and I think that is a better place to put the processing, but I don't think its a particulaly good way to structure managing the memory of the data.

As I was writing up the `DacStreams` contract for the pointer to name mapping, I think I found a better approach.

1. Use the `ProcessedData`/`IData` interfaces to hold the data. I dislike the `RuntimeTypeSystem` approach of having some state in the actual contract implementation. I want ALL state for the contracts to be held in `ProcessedData`, or on other structures associated with the Target (such as the `Metadata`).
2. Put the data structure into the contract implementation.

This both lets us have a single part of the system to do all memory invalidation, lets us get the caching behavior to avoid reprocessing the world all of the time, and places all of the logic that understands runtime data structures into the contract.

lambdageek commented on src/native/managed/cdacreader/src/Data/MethodTableArray.cs:
```diff
@@ -0,0 +1,33 @@
+
+        for (int i = 0; i < key.size; i++)
+        {
+            instantiationSpan[i] = target.Contracts.RuntimeTypeSystem.GetMethodTableHandle(targetPointerSpan[i]);
```
this dependency on the contracts from a data accessor feels wrong.  I think we should have clean separation

lambdageek commented on src/native/managed/cdacreader/src/Data/TypeHandleArray.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace Microsoft.Diagnostics.DataContractReader.Data;
+
+internal sealed class TypeHandleArray : IData<TypeHandleArray, (TargetPointer ptr, int size)>
```
Same here. It feels like this belongs together with the RuntimeTypeSystem contract.

Maybe we need for the Target to provide some kind of caching service for contracts' auxiliary data?  But it seems like it should be something disctinct from the mechanism that reads/caches target process memory.

davidwrighton replied:
It all feels like `ProcessedData` to me. We could have 2 different objects on the target, but I don't think that buys us anything. What I think would be useful is to move these sorts of Data things from being held in the Data directory, to being contained within the file that implements the contract.

lambdageek reviewed (commented):
I didn't look at the EcmaMetadataReader or the TypeNameBuilder yet.  Initial comments to answer questions and raise some concerns about MethodTableArray and TypeHandleArray.

davidwrighton commented on docs/design/datacontracts/RuntimeTypeSystem.md:
```diff
@@ -221,4 +321,194 @@ The contract additionally depends on the `EEClass` data descriptor.
+            CorElementType elemType = (CorElementType)(typeDesc.TypeAndFlags & 0xFF);
+            switch (elemType)
+            {
+                case CorElementType.ValueType:
+                case CorElementType.Byref:
+                case CorElementType.Ptr:
+                    return true;
```
No, I think this is for representing the "pinvoke" view of a valuetype. These aren't officially part of the ECMA world at ALL, but they exist within the CLR runtime (or did at one point in time). I'm not sure that there is any way to get to this code from SOS, but... I'm not 100% certain on that.

davidwrighton replied:
Yep, confirmed. This what the runtime knows as a Native Valuetype. 

See https://github.com/dotnet/runtime/blob/ad25cd0126a8d4a060f2039f987bd58e7ca56a1d/src/coreclr/vm/typehandle.cpp#L262-L265

davidwrighton commented on docs/design/datacontracts/RuntimeTypeSystem.md:
```diff
@@ -154,12 +227,39 @@ internal struct MethodTable_1
-The contract additionally depends on the `EEClass` data descriptor.
+The contract additionally depends on these data descriptors
+
+| Data Descriptor Name |
+| --- |
+| `EEClass` |
+| `ArraayClass` |
+| `TypeDesc` |
+| `ParamTypeDesc` |
+| `TypeVarTypeDesc` |
+| `FnPtrTypeDesc` |
+| `GenericsDictInfo` |
+
```
In my second pass at this today, I went and did a bit where I put all the types, and all the fields, and comments for the meaning of the fields. It looks really nice, and you guys will see it next week.

davidwrighton replied:
The other thing we should probably do is put all the magic constants, like enum values in a similar table. Then we can add some build time validation that a various constants have the expected values for the particular contract version we're working with (This could probably even be _DEBUG only validation if its easier to do.). I don't think we need a data descriptor for those values... just enough validation that we know to update the contract version when magic numbers change. I'm not going to do that in this change, but it seems like a good idea for us to implement soonish.

davidwrighton committed:
Add support for the EEName DacStream

davidwrighton committed:
- Refactor to expose TypeHandle exclusively from the contracts
- Remove TypeHandleArray and MethodTableArray in favor of contract specific logic

davidwrighton committed:
Address concern around data structure for DynamicMetadata

davidwrighton committed:
Merge branch 'cdac-methodtable-name' of github.com:davidwrighton/runtime into cdac-methodtable-name

davidwrighton committed:
Fix build break

davidwrighton committed:
Fix musl build issue

lambdageek commented:
I expected to see updates to the tests due to the `MethodTable`=>`TypeHandle` change.

(VS Code's test extension seems to detect and run the testsuite correctly (at least if you build Debug libraries) if you want to run them locally if you open the test project)

davidwrighton commented on docs/design/datacontracts/RuntimeTypeSystem.md:
```diff
@@ -4,46 +4,80 @@ This contract is for exploring the properties of the runtime types of values on
+{
+    // Values defined in ECMA-335 - II.23.1.16 Element types used in signatures
+    // + 
+    Internal = 0x21, // Indicates that the next pointer sized number of bytes is the address of a TypeHandle. Signatures that contain the Internal CorElementType cannot exist in metadata separate from a live process.
```
Oh, sorry.. I meant to express that they can't be in the disk format or pulled anywhere from a saved metadata block. (Also this detail won't be used until we implement `GetMethodDescName`, but since I was adding CorElementType, I added it at this time.)

lambdageek committed:
update tests

lambdageek commented:
@davidwrighton here's a commit that has the minimal set of changes to make the existing tests pass again (`MethodTableHandle` => `TypeHandle` and also add mock fields for MethodTable and EEClass)

davidwrighton commented on docs/design/datacontracts/DacStreams.md:
```diff
@@ -0,0 +1,81 @@
+        return null;
+    }
+
+    if (_target.Read<uint>(miniMetaDataBuffAddress) != 0x6d727473)
```
From what I can tell, this is the first documentation that anyone has ever made for this format. I'll ... make it a bit nicer.

davidwrighton replied:
The rewrite of the documentation ended up converting the entire implementation of this to prose + tables. For this particular scenario, it looks better than the C#, so I've just replaced the C# with a couple of comments.

davidwrighton committed:
Remove unnecessary usings

davidwrighton committed:
- Transform magic numbers from DacStreams implementation into constants
- Add test suite for the DacStreams contract

davidwrighton committed:
Feedback

davidwrighton commented on src/native/managed/cdacreader/src/Helpers/EcmaMetadataReader.cs:
```diff
@@ -0,0 +1,399 @@
+
+        for (ushort iStream = 0; iStream < streams; iStream++)
+        {
+            var stream = ReadStream(ref image);
+            if (stream.Name == "#Strings")
```
Yeah, probably, but as I've said above... the actual `EcmaMetadataReader` isn't really intended to live for very long as  code, and the current code is ... good enough.
